File: tests/WebAppDemoCsharp.Tests/TodoApiTests.cs
using System.Net;
using System.Net.Http.Json;

using Microsoft.AspNetCore.Mvc.Testing;

namespace WebAppDemoCsharp.Tests;

public class TodoApiTests(WebApplicationFactory<Program> factory) : IClassFixture<WebApplicationFactory<Program>>
{
  private readonly HttpClient _client = factory.CreateClient();

  [Fact]
  public async Task CreateTodoItem_ShouldReturnCreated()
  {
    // Arrange
    var request = new { Title = "Test Todo", Details = "Test Details" };

    // Act
    var response = await _client.PutAsJsonAsync("/api/todo", request);

    // Assert
    Assert.Equal(HttpStatusCode.Created, response.StatusCode);
    var responseBody = await response.Content.ReadFromJsonAsync<TodoItem>();
    Assert.True(Guid.TryParse(responseBody.Id.Value.ToString(), out _),
      "Response does not contain a valid GUID.");
  }
}

File: tests/WebAppDemoFsharo.Tests/TodoApiTests.fs
module TodoApiTests

open System
open System.Net
open System.Net.Http.Json
open Microsoft.AspNetCore.Mvc.Testing
open Xunit
open FsCheck
open FsCheck.Xunit

type Request = { Title: string; Details: string }

[<Fact>]
let ``create todo item should return created`` () =
  let tryParseGuid (input: string) =
    match Guid.TryParse(input) with
    | true, parsedGui -> Some parsedGui
    | false, _ -> None

  task {
    // Arrange
    use factory = new WebApplicationFactory<Program>()
    use client = factory.CreateClient()

    let url = "/api/todo"

    let request = {
      Title = "title1"
      Details = "details1"
    }

    // Act
    let! response = client.PutAsJsonAsync(url, request)

    // Assert
    Assert.True(response.IsSuccessStatusCode, $"Expected success status code but got %A{response.StatusCode}")
    Assert.Equal(HttpStatusCode.Created, response.StatusCode)

    let! responseBody = response.Content.ReadFromJsonAsync<TodoItem>()
    let unparsedId = responseBody.Id.Value.ToString()
    let maybeId = tryParseGuid unparsedId
    Assert.True(maybeId.IsSome, "Expected Some but got None.")
  }

[<Property(Skip = "Does not work")>]
let ``create todo item should handle valid input`` (t: NonWhiteSpaceString, d: NonWhiteSpaceString) =
  let title = t.Get
  let details = d.Get

  let tryParseGuid (input: string) =
    match Guid.TryParse(input) with
    | true, parsedGuid -> Some parsedGuid
    | false, _ -> None

  task {
    // Arrange
    use factory = new WebApplicationFactory<Program>()
    use client = factory.CreateClient()

    let url = "/api/todo"
    let request = { Title = title; Details = details }

    // Act
    let! response = client.PutAsJsonAsync(url, request)

    // Assert
    Assert.True(response.IsSuccessStatusCode, $"Expected success status code but got %A{response.StatusCode}")
    Assert.Equal(HttpStatusCode.Created, response.StatusCode)

    let! responseBody = response.Content.ReadFromJsonAsync<TodoItem>()
    let unparsedId = responseBody.Id.Value.ToString()
    let maybeId = tryParseGuid unparsedId
    Assert.True(maybeId.IsSome, "Expected Some but got None.")
  }